File: src/rohrpost/sync.py
"""The sync round: bidirectional three-way merge per linked ticket (spec §8.4).

For each ticket bound to a remote, sync pulls the live remote item, three-way
merges it against the shadow (merge base) and the folded ticket, applies the
remote-won fields locally (``set`` events, actor ``remote/<name>``), flags
conflicts per policy, pushes the local-won fields to the remote, and rewrites
the shadow from the post-sync remote state.

``body`` gets a genuine three-way text merge (§8.3), while ``labels`` uses
set-wise three-way merge semantics. A crash between applying remote-won locally
and rewriting the shadow leaves a stale shadow (a redundant, idempotent merge
next round) rather than a lost update (§8.4).
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path

from rohrpost import shadow, store
from rohrpost.config import Config
from rohrpost.events import SYNC_TICKET
from rohrpost.exceptions import RemoteItemNotFoundError
from rohrpost.fold import Ticket, fold_all, load_tickets
from rohrpost.merge import FieldConflict, MergeResult, Policy, three_way
from rohrpost.providers import Provider
from rohrpost.util import Clock, now_ts

#: Ticket fields supported by the provider mapping and merge engine.
SYNCED_FIELDS: tuple[str, ...] = ("title", "body", "status", "priority", "labels")


@dataclass(frozen=True, slots=True)
class TicketSync:
    """Per-ticket outcome of a sync round."""

    ticket: str
    ref: str
    pulled: int
    pushed: int
    conflicts: list[str]
    changed: bool = False


@dataclass(frozen=True, slots=True)
class SyncReport:
    """Aggregate outcome of a sync round over one remote."""

    remote: str
    tickets: list[TicketSync] = field(default_factory=list)

    @property
    def pulled(self) -> int:
        return sum(t.pulled for t in self.tickets)

    @property
    def pushed(self) -> int:
        return sum(t.pushed for t in self.tickets)

    @property
    def conflicts(self) -> int:
        return sum(len(t.conflicts) for t in self.tickets)


def _mapped_fields(remote_config: dict[str, object]) -> set[str]:
    """The local field names this remote maps (keys of its ``[fields]`` table)."""
    fields = remote_config.get("fields", {})
    return set(fields.keys()) if isinstance(fields, dict) else set()


def _config_for(remote: str, config: Config) -> dict[str, object]:
    raw = config.remotes.get(remote)
    if raw is None:
        from rohrpost.exceptions import ConfigError

        raise ConfigError(f"no [remotes.{remote}] configured")
    return raw


def _policy_of(remote_config: dict[str, object]) -> Policy:
    p = remote_config.get("policy", "flag")
    if p == "local":
        return "local"
    if p == "remote":
        return "remote"
    return "flag"


def sync_round(
    rohrpost_dir: Path,
    remote: str,
    provider: Provider,
    config: Config,
    *,
    dry_run: bool = False,
    actor: str | None = None,
    now: Clock = now_ts,
    ulid: Callable[[], str] | None = None,
) -> SyncReport:
    """Run one sync round against ``remote``. Returns a :class:`SyncReport`."""
    from rohrpost.events import Event
    from rohrpost.ids import new_ulid

    remote_config = _config_for(remote, config)
    policy = _policy_of(remote_config)
    who = actor or f"remote/{remote}"
    mk_ulid = ulid or new_ulid
    report = SyncReport(remote=remote)

    by_id = load_tickets(rohrpost_dir)
    linked = [(tid, t) for tid, t in by_id.items() if remote in t.remotes]
    ctx = _SyncCtx(rohrpost_dir, remote, who, now, mk_ulid, Event, dry_run)
    mapped = _mapped_fields(remote_config)

    for tid, ticket in linked:
        report.tickets.append(
            _sync_ticket(ctx, tid, ticket, provider, remote_config, policy, mapped)
        )
    if not dry_run and any(ticket.changed for ticket in report.tickets):
        _append_synced(rohrpost_dir, remote, who, now, mk_ulid, Event)
    return report


@dataclass(frozen=True, slots=True)
class _SyncCtx:
    """Bundle of the per-round immutable context handed to the per-ticket helpers."""

    repo: Path
    remote: str
    who: str
    now: Clock
    ulid: Callable[[], str]
    event_cls: type
    dry_run: bool


def _sync_ticket(
    ctx: _SyncCtx,
    tid: str,
    ticket: Ticket,
    provider: Provider,
    remote_config: dict[str, object],
    policy: Policy,
    mapped: set[str],
) -> TicketSync:
    """Fetch, merge, and apply one linked ticket, including deletion handling."""
    ref = ticket.remotes[ctx.remote]
    try:
        merged, live, base, had_shadow = _merge_ticket(ctx, ticket, provider, remote_config, policy)
    except RemoteItemNotFoundError:
        changed = _flag_deleted(ctx, tid, ref, ticket)
        return TicketSync(tid, ref, 0, 0, ["remote"], changed)
    pulled, pushed, changed = _apply_merge(
        ctx, tid, ref, ticket, merged, live, base, had_shadow, provider, mapped
    )
    return TicketSync(
        tid,
        ref,
        pulled,
        pushed,
        [conflict.field for conflict in merged.conflicts],
        changed,
    )


def _merge_ticket(
    ctx: _SyncCtx,
    ticket: Ticket,
    provider: Provider,
    remote_config: dict[str, object],
    policy: Policy,
) -> tuple[MergeResult, dict[str, object], dict[str, object], bool]:
    """Fetch the live remote, load the shadow, and three-way merge the ticket."""
    ref = getattr(ticket, "remotes", {}).get(ctx.remote, "")
    mapped = _mapped_fields(remote_config)
    live = {k: v for k, v in provider.fetch(ref).items() if k in mapped}
    stored = shadow.read_shadow(ctx.repo, ctx.remote, ref)
    base = {k: v for k, v in (stored or {}).items() if k in mapped}
    local = {k: v for k, v in _ticket_fields(ticket).items() if k in mapped}
    if stored is None:
        # A missing/corrupt shadow is not enough information to choose a winner.
        # Establish a base without touching either side; the next genuine edit is
        # then classified normally.
        return MergeResult(), live, base, False
    return three_way(base, local, live, policy=policy), live, base, True


def _apply_merge(
    ctx: _SyncCtx,
    tid: str,
    ref: str,
    ticket: Ticket,
    merged: MergeResult,
    live: dict[str, object],
    base: dict[str, object],
    had_shadow: bool,
    provider: Provider,
    mapped: set[str],
) -> tuple[int, int, bool]:
    """Apply remote-won / conflicts / local-won, push, rewrite shadow. Returns (pulled, pushed)."""
    if _should_defer(ctx, tid, ticket, mapped, had_shadow, live, base):
        return 0, 0, False
    pulled = len(merged.remote_won)
    pushed = len(merged.local_won)
    if ctx.dry_run:
        return pulled, pushed, _has_planned_change(merged, live, base, had_shadow)
    changed = _apply_local_merge(ctx, tid, ticket, merged)
    live, pushed_change = _push_merge(provider, ref, live, merged.local_won)
    shadow_change = _update_shadow(ctx, ref, live, base, had_shadow)
    changed = changed or pushed_change or shadow_change
    return pulled, pushed, changed


def _should_defer(
    ctx: _SyncCtx,
    tid: str,
    ticket: Ticket,
    mapped: set[str],
    had_shadow: bool,
    live: dict[str, object],
    base: dict[str, object],
) -> bool:
    completed_conflict = had_shadow and _is_flagged(ticket, ctx.remote) and live == base
    return completed_conflict or _changed_during_fetch(ctx.repo, tid, ticket, mapped)


def _has_planned_change(
    merged: MergeResult,
    live: dict[str, object],
    base: dict[str, object],
    had_shadow: bool,
) -> bool:
    return bool(
        merged.remote_won
        or merged.local_won
        or merged.conflicts
        or merged.resolved
        or not had_shadow
        or live != base
    )


def _apply_local_merge(ctx: _SyncCtx, tid: str, ticket: Ticket, merged: MergeResult) -> bool:
    if merged.conflicts:
        changed = _flag_conflict(ctx, tid, ticket, merged.conflicts, merged.remote_won)
    else:
        changed = bool(merged.remote_won) and _append_set(ctx, tid, ticket, merged.remote_won)
    for conflict in merged.resolved:
        changed = _record_resolution(ctx, tid, ticket, conflict, merged.local_won) or changed
    return changed


def _push_merge(
    provider: Provider,
    ref: str,
    live: dict[str, object],
    local_won: dict[str, object],
) -> tuple[dict[str, object], bool]:
    if not local_won:
        return live, False
    return provider.push(ref, dict(local_won)), True


def _update_shadow(
    ctx: _SyncCtx,
    ref: str,
    live: dict[str, object],
    base: dict[str, object],
    had_shadow: bool,
) -> bool:
    next_shadow = dict(live)
    if had_shadow and next_shadow == base:
        return False
    shadow.write_shadow(ctx.repo, ctx.remote, ref, next_shadow)
    return True


def _ticket_fields(ticket: Ticket) -> dict[str, object]:
    """Extract fields understood by the sync merge engine from a ticket."""
    return {name: getattr(ticket, name) for name in SYNCED_FIELDS}


def _append_set(
    ctx: _SyncCtx,
    tid: str,
    ticket: Ticket,
    fields: dict[str, object],
) -> bool:
    """Apply remote-won fields locally as one provenance-stamped event."""
    payload = _event_payload(ticket, fields)
    if not payload:
        return False
    store.append_event(
        ctx.repo,
        ctx.event_cls(
            id=ctx.ulid(),
            ts=ctx.now(),
            ticket=tid,
            op="set",
            actor=ctx.who,
            set=payload,
        ),
    )
    return True


def _event_payload(ticket: Ticket, fields: dict[str, object]) -> dict[str, object]:
    """Translate merged whole values to the event log's scalar/set operations."""
    payload: dict[str, object] = {}
    for name, value in fields.items():
        if name != "labels":
            if getattr(ticket, name) != value:
                payload[name] = value
            continue
        current = set(ticket.labels)
        target = {str(item) for item in value} if isinstance(value, list) else set()
        added = sorted(target - current)
        removed = sorted(current - target)
        if added:
            payload["labels+"] = added
        if removed:
            payload["labels-"] = removed
    return payload


def _flag_conflict(
    ctx: _SyncCtx,
    tid: str,
    ticket: Ticket,
    conflicts: list[FieldConflict],
    inbound: dict[str, object],
) -> bool:
    """Apply inbound values/markers and the conflict flag atomically."""
    detail = "; ".join(f"{c.field}: local={c.local!r} remote={c.remote!r}" for c in conflicts)
    comment = f"sync conflict with {ctx.remote} — {detail}"
    changed = _append_comment_once(ctx, tid, ticket, comment)
    updates = dict(inbound)
    updates["status"] = "review"
    labels = updates.get("labels", ticket.labels)
    target_labels = (
        {str(item) for item in labels} if isinstance(labels, list) else set(ticket.labels)
    )
    target_labels.add(f"conflict:{ctx.remote}")
    updates["labels"] = sorted(target_labels)
    return _append_set(ctx, tid, ticket, updates) or changed


def _record_resolution(
    ctx: _SyncCtx,
    tid: str,
    ticket: Ticket,
    conflict: FieldConflict,
    local_won: dict[str, object],
) -> bool:
    winner = "local" if conflict.field in local_won else "remote"
    text = (
        f"sync conflict with {ctx.remote} resolved by {winner} policy — "
        f"{conflict.field}: local={conflict.local!r} remote={conflict.remote!r}"
    )
    return _append_comment_once(ctx, tid, ticket, text)


def _append_comment_once(ctx: _SyncCtx, tid: str, ticket: Ticket, text: str) -> bool:
    if any(comment.actor == ctx.who and comment.text == text for comment in ticket.comments):
        return False
    store.append_event(
        ctx.repo,
        ctx.event_cls(
            id=ctx.ulid(),
            ts=ctx.now(),
            ticket=tid,
            op="comment",
            actor=ctx.who,
            text=text,
        ),
    )
    return True


def _flag_deleted(ctx: _SyncCtx, tid: str, ref: str, ticket: Ticket) -> bool:
    conflict = FieldConflict("remote", f"linked ticket {tid}", f"missing item {ref}")
    if ctx.dry_run:
        detail = f"remote: local={conflict.local!r} remote={conflict.remote!r}"
        text = f"sync conflict with {ctx.remote} — {detail}"
        has_comment = any(c.actor == ctx.who and c.text == text for c in ticket.comments)
        return not has_comment or not _is_flagged(ticket, ctx.remote) or ticket.status != "review"
    return _flag_conflict(ctx, tid, ticket, [conflict], {})


def _is_flagged(ticket: Ticket, remote: str) -> bool:
    return f"conflict:{remote}" in ticket.labels


def _changed_during_fetch(repo: Path, tid: str, before: Ticket, mapped: set[str]) -> bool:
    current = fold_all(repo).get(tid)
    if current is None:
        return True
    before_fields = {k: v for k, v in _ticket_fields(before).items() if k in mapped}
    current_fields = {k: v for k, v in _ticket_fields(current).items() if k in mapped}
    return current_fields != before_fields


def _append_synced(
    rohrpost_dir: Path,
    remote: str,
    who: str,
    now: Clock,
    ulid: Callable[[], str],
    event_cls: type,
) -> None:
    """Record that a ticket completed a sync round with ``remote`` (§8.4 step 6)."""
    timestamp = now()
    store.append_event(
        rohrpost_dir,
        event_cls(
            id=ulid(),
            ts=timestamp,
            ticket=SYNC_TICKET,
            op="synced",
            actor=who,
            remote=remote,
            at=timestamp,
        ),
    )


__all__ = ["SYNCED_FIELDS", "SyncReport", "TicketSync", "sync_round"]

File: cpp/src/sync.cpp
#include "rohrpost/sync.hpp"

#include "rohrpost/errors.hpp"
#include "rohrpost/merge.hpp"
#include "rohrpost/shadow.hpp"
#include "rohrpost/store.hpp"

#include <algorithm>
#include <format>
#include <set>

namespace rp::sync {
namespace {

namespace fs = std::filesystem;

int SyncReport_sum(const std::vector<TicketSync>& tickets, int TicketSync::*member) {
    int total = 0;
    for (const auto& t : tickets) total += t.*member;
    return total;
}

std::set<std::string> mapped_fields(const Json& remote_config) {
    std::set<std::string> out;
    const auto it = remote_config.find("fields");
    if (it != remote_config.end() && it->is_object()) {
        for (const auto& [k, v] : it->items()) out.insert(k);
    }
    return out;
}

merge::Policy policy_of(const Json& remote_config) {
    const auto it = remote_config.find("policy");
    if (it != remote_config.end() && it->is_string()) {
        const std::string p = it->get<std::string>();
        if (p == "local") return merge::Policy::Local;
        if (p == "remote") return merge::Policy::Remote;
    }
    return merge::Policy::Flag;
}

/// Extract the fields the merge engine understands from a ticket.
Json ticket_fields(const Ticket& t) {
    Json out = Json::object();
    out["title"] = t.title;
    out["body"] = t.body ? Json(*t.body) : Json();
    out["status"] = t.status;
    out["priority"] = t.priority;
    out["labels"] = t.labels;
    return out;
}

Json filter_mapped(const Json& fields, const std::set<std::string>& mapped) {
    Json out = Json::object();
    if (!fields.is_object()) return out;
    for (const auto& [k, v] : fields.items()) {
        if (mapped.contains(k)) out[k] = v;
    }
    return out;
}

struct Ctx {
    fs::path repo;
    std::string remote;
    std::string who;
    const api::Sources& sources;
    bool dry_run;
};

bool is_flagged(const Ticket& t, const std::string& remote) {
    return std::find(t.labels.begin(), t.labels.end(), "conflict:" + remote) != t.labels.end();
}

bool append_comment_once(const Ctx& ctx, const std::string& tid, const Ticket& ticket, const std::string& text) {
    for (const auto& c : ticket.comments) {
        if (c.actor == ctx.who && c.text == text) return false;
    }
    Event e;
    e.id = ctx.sources.ulid();
    e.ts = ctx.sources.now();
    e.ticket = tid;
    e.op = "comment";
    e.actor = ctx.who;
    e.text = text;
    store::append_event(ctx.repo, e);
    return true;
}

/// Translate merged whole values to the event log's scalar/set operations.
Json event_payload(const Ticket& ticket, const Json& fields) {
    Json payload = Json::object();
    const Json current_fields = ticket_fields(ticket);
    for (const auto& [name, value] : fields.items()) {
        if (name != "labels") {
            if (!json::py_equal(current_fields.contains(name) ? current_fields[name] : Json(), value)) payload[name] = value;
            continue;
        }
        const std::set<std::string> current(ticket.labels.begin(), ticket.labels.end());
        std::set<std::string> target;
        if (value.is_array()) {
            for (const auto& item : value) target.insert(json::py_str(item));
        }
        Json added = Json::array();
        Json removed = Json::array();
        for (const auto& l : target) {
            if (!current.contains(l)) added.push_back(l);
        }
        for (const auto& l : current) {
            if (!target.contains(l)) removed.push_back(l);
        }
        if (!added.empty()) payload["labels+"] = added;
        if (!removed.empty()) payload["labels-"] = removed;
    }
    return payload;
}

bool append_set(const Ctx& ctx, const std::string& tid, const Ticket& ticket, const Json& fields) {
    const Json payload = event_payload(ticket, fields);
    if (payload.empty()) return false;
    Event e;
    e.id = ctx.sources.ulid();
    e.ts = ctx.sources.now();
    e.ticket = tid;
    e.op = "set";
    e.actor = ctx.who;
    e.set = payload;
    store::append_event(ctx.repo, e);
    return true;
}

std::string conflict_detail(const std::vector<merge::FieldConflict>& conflicts) {
    std::string detail;
    for (std::size_t i = 0; i < conflicts.size(); ++i) {
        if (i) detail += "; ";
        detail += std::format("{}: local={} remote={}", conflicts[i].field, json::py_repr(conflicts[i].local), json::py_repr(conflicts[i].remote));
    }
    return detail;
}

bool flag_conflict(const Ctx& ctx, const std::string& tid, const Ticket& ticket,
                   const std::vector<merge::FieldConflict>& conflicts, const Json& inbound) {
    const std::string comment = std::format("sync conflict with {} — {}", ctx.remote, conflict_detail(conflicts));
    const bool changed = append_comment_once(ctx, tid, ticket, comment);
    Json updates = inbound.is_object() ? inbound : Json::object();
    updates["status"] = "review";
    std::set<std::string> target_labels;
    if (updates.contains("labels") && updates["labels"].is_array()) {
        for (const auto& item : updates["labels"]) target_labels.insert(json::py_str(item));
    } else {
        target_labels.insert(ticket.labels.begin(), ticket.labels.end());
    }
    target_labels.insert("conflict:" + ctx.remote);
    updates["labels"] = std::vector<std::string>(target_labels.begin(), target_labels.end());
    const bool wrote = append_set(ctx, tid, ticket, updates);
    return wrote || changed;
}

bool record_resolution(const Ctx& ctx, const std::string& tid, const Ticket& ticket, const merge::FieldConflict& conflict,
                       const Json& local_won) {
    const char* winner = local_won.contains(conflict.field) ? "local" : "remote";
    const std::string text = std::format("sync conflict with {} resolved by {} policy — {}: local={} remote={}", ctx.remote, winner,
                                         conflict.field, json::py_repr(conflict.local), json::py_repr(conflict.remote));
    return append_comment_once(ctx, tid, ticket, text);
}

bool flag_deleted(const Ctx& ctx, const std::string& tid, const std::string& ref, const Ticket& ticket) {
    const merge::FieldConflict conflict{"remote", Json(std::format("linked ticket {}", tid)), Json(std::format("missing item {}", ref)), std::nullopt};
    if (ctx.dry_run) {
        const std::string text = std::format("sync conflict with {} — remote: local={} remote={}", ctx.remote, json::py_repr(conflict.local), json::py_repr(conflict.remote));
        bool has_comment = false;
        for (const auto& c : ticket.comments) {
            if (c.actor == ctx.who && c.text == text) has_comment = true;
        }
        return !has_comment || !is_flagged(ticket, ctx.remote) || ticket.status != "review";
    }
    return flag_conflict(ctx, tid, ticket, {conflict}, Json::object());
}

bool changed_during_fetch(const fs::path& repo, const std::string& tid, const Ticket& before, const std::set<std::string>& mapped) {
    const TicketMap current_map = fold_all(repo);
    const Ticket* current = current_map.find(tid);
    if (current == nullptr) return true;
    return !json::py_equal(filter_mapped(ticket_fields(*current), mapped), filter_mapped(ticket_fields(before), mapped));
}

struct Merged {
    merge::MergeResult result;
    Json live;
    Json base;
    bool had_shadow;
};

Merged merge_ticket(const Ctx& ctx, const Ticket& ticket, providers::Provider& provider, const Json& remote_config,
                    merge::Policy policy) {
    const std::string* ref_ptr = ticket.remotes.find(ctx.remote);
    const std::string ref = ref_ptr ? *ref_ptr : "";
    const auto mapped = mapped_fields(remote_config);
    const Json live = filter_mapped(provider.fetch(ref), mapped);
    const auto stored = shadow::read_shadow(ctx.repo, ctx.remote, ref);
    const Json base = filter_mapped(stored.value_or(Json::object()), mapped);
    const Json local = filter_mapped(ticket_fields(ticket), mapped);
    if (!stored) {
        // A missing/corrupt shadow is not enough information to choose a
        // winner: establish a base without touching either side.
        return Merged{merge::MergeResult{}, live, base, false};
    }
    return Merged{merge::three_way(base, local, live, policy), live, base, true};
}

bool has_planned_change(const merge::MergeResult& m, const Json& live, const Json& base, bool had_shadow) {
    return !m.remote_won.empty() || !m.local_won.empty() || !m.conflicts.empty() || !m.resolved.empty() ||
           !had_shadow || !json::py_equal(live, base);
}

TicketSync sync_ticket(const Ctx& ctx, const std::string& tid, const Ticket& ticket, providers::Provider& provider,
                       const Json& remote_config, merge::Policy policy, const std::set<std::string>& mapped) {
    const std::string ref = *ticket.remotes.find(ctx.remote);
    Merged merged;
    try {
        merged = merge_ticket(ctx, ticket, provider, remote_config, policy);
    } catch (const RemoteItemNotFoundError&) {
        const bool changed = flag_deleted(ctx, tid, ref, ticket);
        return TicketSync{tid, ref, 0, 0, {"remote"}, changed};
    }
    // _apply_merge
    const bool completed_conflict = merged.had_shadow && is_flagged(ticket, ctx.remote) && json::py_equal(merged.live, merged.base);
    if (completed_conflict || changed_during_fetch(ctx.repo, tid, ticket, mapped)) {
        std::vector<std::string> conflicts;
        for (const auto& c : merged.result.conflicts) conflicts.push_back(c.field);
        return TicketSync{tid, ref, 0, 0, conflicts, false};
    }
    const int pulled = static_cast<int>(merged.result.remote_won.size());
    int pushed = static_cast<int>(merged.result.local_won.size());
    std::vector<std::string> conflicts;
    for (const auto& c : merged.result.conflicts) conflicts.push_back(c.field);
    if (ctx.dry_run) {
        return TicketSync{tid, ref, pulled, pushed, conflicts, has_planned_change(merged.result, merged.live, merged.base, merged.had_shadow)};
    }
    // _apply_local_merge
    bool changed;
    if (!merged.result.conflicts.empty()) {
        changed = flag_conflict(ctx, tid, ticket, merged.result.conflicts, merged.result.remote_won);
    } else {
        changed = !merged.result.remote_won.empty() && append_set(ctx, tid, ticket, merged.result.remote_won);
    }
    for (const auto& conflict : merged.result.resolved) {
        changed = record_resolution(ctx, tid, ticket, conflict, merged.result.local_won) || changed;
    }
    // _push_merge
    Json live = merged.live;
    bool pushed_change = false;
    if (!merged.result.local_won.empty()) {
        live = provider.push(ref, merged.result.local_won);
        pushed_change = true;
    }
    // _update_shadow
    bool shadow_change = false;
    if (!(merged.had_shadow && json::py_equal(live, merged.base))) {
        shadow::write_shadow(ctx.repo, ctx.remote, ref, live);
        shadow_change = true;
    }
    changed = changed || pushed_change || shadow_change;
    return TicketSync{tid, ref, pulled, pushed, conflicts, changed};
}

void append_synced(const fs::path& dir, const std::string& remote, const std::string& who, const api::Sources& s) {
    const std::string timestamp = s.now();
    Event e;
    e.id = s.ulid();
    e.ts = timestamp;
    e.ticket = std::string(kSyncTicket);
    e.op = "synced";
    e.actor = who;
    e.remote = remote;
    e.at = timestamp;
    store::append_event(dir, e);
}

}  // namespace

int SyncReport::pulled() const { return SyncReport_sum(tickets, &TicketSync::pulled); }
int SyncReport::pushed() const { return SyncReport_sum(tickets, &TicketSync::pushed); }
int SyncReport::conflicts() const {
    int total = 0;
    for (const auto& t : tickets) total += static_cast<int>(t.conflicts.size());
    return total;
}

SyncReport sync_round(const fs::path& rohrpost_dir, const std::string& remote, providers::Provider& provider,
                      const Config& config, const SyncOptions& options) {
    const Json* raw = config.remotes.find(remote);
    if (raw == nullptr) throw ConfigError(std::format("no [remotes.{}] configured", remote));
    const Json& remote_config = *raw;
    const merge::Policy policy = policy_of(remote_config);
    const std::string who = options.actor.value_or("remote/" + remote);
    SyncReport report{remote, {}};

    const TicketMap by_id = load_tickets(rohrpost_dir);
    std::vector<std::pair<std::string, Ticket>> linked;
    for (const auto& [tid, t] : by_id) {
        if (t.remotes.contains(remote)) linked.emplace_back(tid, t);
    }
    const Ctx ctx{rohrpost_dir, remote, who, options.sources, options.dry_run};
    const auto mapped = mapped_fields(remote_config);
    for (const auto& [tid, ticket] : linked) {
        report.tickets.push_back(sync_ticket(ctx, tid, ticket, provider, remote_config, policy, mapped));
    }
    const bool any_changed = std::any_of(report.tickets.begin(), report.tickets.end(), [](const TicketSync& t) { return t.changed; });
    if (!options.dry_run && any_changed) append_synced(rohrpost_dir, remote, who, options.sources);
    return report;
}

}  // namespace rp::sync
